src/sysevent.c

```diff
@@ -0,0 +1,984 @@
+    // fine when the process isn't supposed to be exiting, but in
+    // the case of a process shutdown, we don't want to have an
+    // idle thread hanging around.  Calling pthread_cancel here in
+    // the case of a shutdown is just assures that the thread is
+    // gone and that the process has been fully terminated.
+
+    DEBUG("sysevent plugin: Canceling thread for process shutdown");
+
+    status = pthread_cancel(sysevent_thread_id);
+
+    if (status != 0) {
+      ERROR("sysevent plugin: Unable to cancel thread: %d (%s)", status,
+            strerror(errno));
+      status = -1;
+    }
+  } else {
+    status = pthread_join(sysevent_thread_id, /* return = */ NULL);
+    if (status != 0) {
+      ERROR("sysevent plugin: Stopping thread failed.");
+      status = -1;
+    }
+  }
+
+  pthread_mutex_lock(&sysevent_lock);
+  memset(&sysevent_thread_id, 0, sizeof(sysevent_thread_id));
+  sysevent_thread_error = 0;
+  pthread_mutex_unlock(&sysevent_lock);
+
+  DEBUG("sysevent plugin: Finished requesting stop of thread");
+
+  return (status);
+} /* }}} int stop_thread */
+
+static int sysevent_init(void) /* {{{ */
+{
+  ring.head = 0;
+  ring.tail = 0;
+  ring.maxLen = buffer_length;
+  ring.buffer = (char **)malloc(buffer_length * sizeof(char *));
+
+  for (int i = 0; i < buffer_length; i++) {
+    ring.buffer[i] = malloc(listen_buffer_size);
+  }
+
+  ring.timestamp = (long long unsigned int *)malloc(
+      buffer_length * sizeof(long long unsigned int));
+
+  if (sock == -1) {
+    const char *hostname = listen_ip;
+    const char *portname = listen_port;
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_DGRAM;
+    hints.ai_protocol = 0;
+    hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG;
+    struct addrinfo *res = 0;
+
+    int err = getaddrinfo(hostname, portname, &hints, &res);
+
+    if (err != 0) {
+      ERROR("sysevent plugin: failed to resolve local socket address (err=%d)",
+            err);
+      freeaddrinfo(res);
+      return (-1);
+    }
+
+    sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+    if (sock == -1) {
+      ERROR("sysevent plugin: failed to open socket: %s", strerror(errno));
+      freeaddrinfo(res);
+      return (-1);
+    }
+
+    if (bind(sock, res->ai_addr, res->ai_addrlen) == -1) {
+      ERROR("sysevent plugin: failed to bind socket: %s", strerror(errno));
+      freeaddrinfo(res);
+      return (-1);
+    }
+
+    freeaddrinfo(res);
+  }
+
+  DEBUG("sysevent plugin: socket created and bound");
+
+  return (start_thread());
+} /* }}} int sysevent_init */
+
+static int sysevent_config_add_listen(const oconfig_item_t *ci) /* {{{ */
+{
+  if (ci->values_num != 2 || ci->values[0].type != OCONFIG_TYPE_STRING ||
+      ci->values[1].type != OCONFIG_TYPE_STRING) {
+    ERROR("sysevent plugin: The `%s' config option needs "
+          "two string arguments (ip and port).",
+          ci->key);
```

Comment:
can be appended to previous line

Reply:
This break-up was done by the required clang formatting and needs to stay as-is to avoid violating the styling.